include/bout/region.hxx

```diff
@@ -0,0 +1,251 @@
+    for (auto block = blocks.begin(); block < blocks.end(); ++block) {                   \
+      for (auto index = block->first; index <= block->second; ++index) {                 \
+        __VA_ARGS__                                                                      \
+      }                                                                                  \
+    }                                                                                    \
+  }
+
+#define BLOCK_REGION_LOOP(region, index, ...)                                            \
+  {                                                                                      \
+    const auto blocks = region.getBlocks();                                              \
+  BOUT_OMP(parallel for)                                                                 \
+    for (auto block = blocks.begin(); block < blocks.end(); ++block) {                   \
+      for (auto index = block->first; index <= block->second; ++index) {                 \
+        __VA_ARGS__                                                                      \
+      }                                                                                  \
+    }                                                                                    \
+  }
+
+class specificInd {
+public:
+  int ind;
+  specificInd() : ind(-1){};
+  specificInd(int i) : ind(i){};
+  specificInd &operator++() {
+    ++ind;
+    return *this;
+  }
+  specificInd operator++(int) {
+    specificInd original(*this);
+    ++ind;
+    return original;
+  }
+  bool operator==(const specificInd &x) const { return ind == x.ind; }
+  bool operator!=(const specificInd &x) const { return ind != x.ind; }
+  bool operator>(const specificInd &x) const { return ind > x.ind; }
+  bool operator>=(const specificInd &x) const { return ind >= x.ind; }
+  bool operator<(const specificInd &x) const { return ind < x.ind; }
+  bool operator<=(const specificInd &x) const { return ind <= x.ind; }
+};
+// Make identical subclasses with different names
+class ind3D : public specificInd {
+public:
+  ind3D() : specificInd(){};
+  ind3D(int i) : specificInd(i){};
+  // Note operator= from base class is always hidden
+  // by implicit method so have to be explicit
+  ind3D &operator=(int i) {
+    ind = i;
+    return *this;
+  }
+};
+class ind2D : public specificInd {
+public:
+  ind2D() : specificInd(){};
+  ind2D(int i) : specificInd(i){};
+  ind2D &operator=(int i) {
+    ind = i;
+    return *this;
+  }
+};
+
+template <typename T = ind3D> class Region {
+  // Following prevents a Region being created with anything other
+  // than ind2D or ind3D as template type
+  static_assert(std::is_base_of<ind2D, T>::value || std::is_base_of<ind3D, T>::value,
+                "Region must be templated with either ind2D or ind3D");
+
+public:
+  typedef T data_type;
+
+  /// Indices to iterate over
+  typedef std::vector<T> regionIndices;
```

Comment:
typedefs should start with a capital, `RegionIndices`, etc.

Reply:
Sure, although this is not at all consistently adopted throughout the code!

Reply:
Agree we should be consistent (I think as my specificInd started as a struct I hadn't noted this).